File: day22/day22.kt
import java.io.File

typealias Cards = List<Int>
fun play(player1: Cards, player2: Cards, recursive: Boolean): Pair<Cards, Cards> {
    val played = mutableSetOf<List<Int>>()
    val p1 = ArrayDeque<Int>(player1)
    val p2 = ArrayDeque<Int>(player2)
    while (p1.isNotEmpty() && p2.isNotEmpty()) {
        val state = listOf(p1.size) + p1 + p2
        if (played.contains(state)) break
        val c1 = p1.removeFirst()
        val c2 = p2.removeFirst()
        var win = if (recursive && p1.size >= c1 && p2.size >= c2) {
            val (t1, _) = play(p1.take(c1), p2.take(c2), true)
            t1.size > 0
        } else c1 > c2
        if (win) p1.addAll(listOf(c1, c2)) else p2.addAll(listOf(c2, c1))
        played.add(state)
    }
    return Pair(p1.toList(), p2.toList())
}

fun score(res: Pair<Cards, Cards>) = 
    (if (res.first.size > 0) res.first else res.second).reversed()
    .mapIndexed { k, v -> v * (k + 1) }.sum()

fun main() {
    val input = File("day22.txt").readText().trimEnd()
    val (player1, player2) = input.split("\n\n").map {
        it.split('\n').drop(1).map { it.toInt() }
    }
    println(score(play(player1, player2, false)))
    println(score(play(player1, player2, true)))
}

File: day22/day22.py
from collections import deque

input_ = open("day22.txt").read().rstrip()
player1, player2 = [list(map(int, x.split('\n')[1:]))
                    for x in input_.split('\n\n')]

def play(player1, player2, recursive):
    played = set()
    p1, p2 = deque(player1), deque(player2)
    while p1 and p2:
        state = (len(p1),) + tuple(p1 + p2)
        if state in played: break
        c1, c2 = p1.popleft(), p2.popleft()
        if recursive and len(p1) >= c1 and len(p2) >= c2:
            s1, s2 = list(p1)[:c1], list(p2)[:c2]
            t1, t2 = play(s1, s2, True)
            win = len(t1) > 0
        else:
            win = c1 > c2
        if win:
            p1.extend((c1, c2))
        else:
            p2.extend((c2, c1))
        played.add(state)
    return p1, p2

score = lambda r: sum(v * (k + 1) for k, v in enumerate(reversed(r[0] or r[1])))
print(score(play(player1, player2, False)))
print(score(play(player1, player2, True)))
